src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MinecraftJarConfiguration.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * This file is part of fabric-loom, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) 2022 FabricMC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package net.fabricmc.loom.configuration.providers.minecraft;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import org.gradle.api.Project;
+
+import net.fabricmc.loom.configuration.decompile.DecompileConfiguration;
+import net.fabricmc.loom.configuration.decompile.SingleJarDecompileConfiguration;
+import net.fabricmc.loom.configuration.decompile.SplitDecompileConfiguration;
+import net.fabricmc.loom.configuration.processors.JarProcessorManager;
+import net.fabricmc.loom.configuration.providers.minecraft.mapped.IntermediaryMinecraftProvider;
+import net.fabricmc.loom.configuration.providers.minecraft.mapped.MappedMinecraftProvider;
+import net.fabricmc.loom.configuration.providers.minecraft.mapped.NamedMinecraftProvider;
+import net.fabricmc.loom.configuration.providers.minecraft.mapped.ProcessedNamedMinecraftProvider;
+
+public enum MinecraftJarConfiguration {
+	MERGED(
+		MergedMinecraftProvider::new,
+		IntermediaryMinecraftProvider.MergedImpl::new,
+		NamedMinecraftProvider.MergedImpl::new,
+		ProcessedNamedMinecraftProvider.MergedImpl::new,
+		SingleJarDecompileConfiguration.Merged::new,
+		List.of("client", "server")
+	),
+	SERVER_ONLY(
+		ServerOnlyMinecraftProvider::new,
+		IntermediaryMinecraftProvider.ServerOnlyImpl::new,
+		NamedMinecraftProvider.ServerOnlyImpl::new,
+		ProcessedNamedMinecraftProvider.ServerOnlyImpl::new,
+		SingleJarDecompileConfiguration.ServerOnly::new,
+		List.of("server")
+	),
```

Comment:
I'm very doubtful that there needs to be yet another set of these minecraft providers. From this class I can see references to 12 separate MinecraftProvider implementation classes, in addition to the different minecraft provider base types for different mapping types, which really feels like bad design.

Reply:
Yes, I somewhat agree with you, im not super happy about the amount of classes and the mess of generics just bellow this. However having everything split out does make it easier to see whats going and less brittle. Adding any more would make the class count explode however I cannot think of any more that might be needed. Moving their construction all into this single location vastly helps as well.

I will have another pass at it to try and reduce the amount of classes, but I dont really fancy rewriting it all again just now.